modules/runner/src/main/java/org/apache/ignite/internal/configuration/storage/LocalFileConfigurationStorage.java

```diff
@@ -0,0 +1,224 @@
+import static org.apache.ignite.internal.configuration.util.ConfigurationUtil.wrongTypeException;
+
+import com.typesafe.config.Config;
+import com.typesafe.config.ConfigFactory;
+import com.typesafe.config.ConfigList;
+import com.typesafe.config.ConfigParseOptions;
+import com.typesafe.config.ConfigRenderOptions;
+import com.typesafe.config.ConfigValue;
+import com.typesafe.config.ConfigValueType;
+import java.io.IOException;
+import java.io.Serializable;
+import java.lang.reflect.Array;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
+import org.apache.ignite.configuration.annotation.ConfigurationType;
+import org.apache.ignite.internal.configuration.NodeBootstrapConfiguration;
+import org.apache.ignite.internal.configuration.NodeConfigCreateException;
+import org.apache.ignite.internal.configuration.NodeConfigWriteException;
+import org.apache.ignite.internal.future.InFlightFutures;
+import org.apache.ignite.internal.logger.IgniteLogger;
+import org.apache.ignite.internal.logger.Loggers;
+import org.apache.ignite.internal.thread.NamedThreadFactory;
+import org.apache.ignite.internal.util.IgniteUtils;
+
+/**
+ * Implementation of {@link ConfigurationStorage} based on local file configuration storage.
+ */
+public class LocalFileConfigurationStorage implements ConfigurationStorage {
+    private static final IgniteLogger LOG = Loggers.forClass(LocalFileConfigurationStorage.class);
+
+    private final Path configPath;
+    private final Map<String, Serializable> lastState = new HashMap<>();
+
+    private final AtomicReference<ConfigurationStorageListener> lsnrRef = new AtomicReference<>();
+
+    private final ExecutorService threadPool = Executors.newFixedThreadPool(4, new NamedThreadFactory("loc-cfg-file", LOG));
+
+    private final InFlightFutures futureTracker = new InFlightFutures();
+
+    public LocalFileConfigurationStorage(NodeBootstrapConfiguration configuration) {
+        this.configPath = configuration.config();
+        ensureConfigFile();
+    }
+
+    @Override
+    public CompletableFuture<Data> readDataOnRecovery() {
+        return readAllLatest("").thenApply(stringMap -> new Data(stringMap, 0));
+    }
+
+    @Override
+    public CompletableFuture<Map<String, ? extends Serializable>> readAllLatest(String prefix) {
+        ensureConfigFile();
+        Map<String, Serializable> map = new HashMap<>();
+        if (!lastState.isEmpty()) {
+            for (Entry<String, ConfigValue> e : parseConfigOptions().entrySet()) {
```

Comment:
There must be a huge TODO somewhere, telling that current code will break once we introduce a named list configuration to the set of node configurations.